src/text/ion_data_source.rs

```diff
@@ -0,0 +1,67 @@
+use std::fs::File;
+use std::io;
+use std::io::{BufRead, BufReader, Read};
+
+/// Types that implement this trait can be converted into an implementation of [io::BufRead],
+/// allowing users to build a [Reader] from a variety of types that might not define I/O operations
+/// on their own.
+pub trait ToIonDataSource {
```

Comment:
🗺️ This trait was previously called `TextIonDataSource`, but it's actually useful for any format. I've renamed it to be more general.

A `Reader` can be built from any type that implements `ToIonDataSource`. I can call `ReaderBuilder::new().build(...)` with a `&str` or `String` of Ion text, a `Vec` of bytes, a file to read from, etc.